File: prefix_sum/shortest_subarray.py

def find_shortest_subarray(integers: list[int], target: int):  # LeetCode Q.862.
    """Find the shortest subarray with sum >= target."""
    prefix_sum, min_len = 0, float("inf")
    queue = []  # Prefix sum end idx increasing monotonic queue.
    for idx, integer in enumerate(integers):
        if integer >= target:  # Base case.
            return 1

        if idx == 0:  # 1st int.
            prefix_sum += integer
            queue.append((prefix_sum, 0))  # Format: (prefix sum, its ending idx).
            continue

        prefix_sum += integer  # Update prefix sum ending at idx.

        # Subarrays from (queue[0][1] + 1)th idx to (idx)th idx.
        while queue and prefix_sum - queue[0][0] >= target:
            if idx - queue[0][1] < min_len:
                min_len = idx - queue[0][1]
            queue.pop(0)

        if prefix_sum >= target and idx + 1 < min_len:  # Subarray from 0th idx to (idx)th idx.
            min_len = idx + 1

        # Make queue's prefix sums smaller as bigger subarray sum is preferred.
        while queue and prefix_sum < queue[-1][0]:
            queue.pop(-1)

        queue.append((prefix_sum, idx))  # Format: (prefix sum, its ending idx).

    return min_len if min_len != float("inf") else -1

File: queue/shortest_subarray/shortest_subarray.js
var find_shortest_subarray_length = function (nums, k) { // LeetCode Q.862.
    let queue = []; // Format: [idx, prefix sum].
    let queue_front_idx = 0; // Locates the current front element.
    let prefix_sum = 0, min_len = nums.length + 1;

    for (const [idx, num] of nums.entries()) {
        prefix_sum += num;
        while (queue_front_idx < queue.length) {
            const [past_idx, past_prefix_sum] = queue.at(queue_front_idx);
            if (prefix_sum - past_prefix_sum < k) {
                break;
            }

            let subarray_len = idx - past_idx;
            if (subarray_len < min_len)
                min_len = subarray_len;
            queue_front_idx++;
        }

        while (queue.length > queue_front_idx && queue.at(-1)[1] >= prefix_sum)
            queue.pop();

        if (prefix_sum >= k && idx + 1 < min_len)
            min_len = idx + 1;

        queue.push([idx, prefix_sum]);
    }

    return (min_len === nums.length + 1) ? -1 : min_len;
};